File: Data Structures & Algorithms/maximum-product-subarray/submission-3.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        max_product=min_product=nums[0]
        for i in range(1,len(nums)):
            prev_min=min_product
            prev_max=max_product
            min_product=min(prev_min*nums[i],prev_max*nums[i],nums[i])
            max_product=max(prev_min*nums[i],prev_max*nums[i],nums[i])
            nums[i]=max(min_product,max_product)
            #print(nums[i],min_product,max_product)
        return max(nums)

File: Data Structures & Algorithms/jump-game/submission-3.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
       vector<bool>reachable(nums.size(),false);
       int n=nums.size();
       reachable[n-1]=true;

       for(int i=n-2;i>=0;i--) {
            int maxAchievedInd=min(n-1,i+nums[i]);
            for(int j=i+1;j<=maxAchievedInd;j++) {
                if(reachable[j]==true) {
                    reachable[i]=true;
                    break;
                }
                
            }
            //cout<<i<<" "<<reachable[i]<<endl;
       } 

       return reachable[0];
    }
};
